src/main/java/teammates/storage/search/SearchManager.java

```diff
@@ -44,108 +47,157 @@ private SearchManager() {
      * Creates or updates the search document for the given document and index.
      */
     public static void putDocument(String indexName, Document document) {
-        Index index = getIndex(indexName);
-
-        int delay = 2;
-        for (int attempts = 0; attempts < MAX_RETRIES; attempts++) {
-            try {
-                PutResponse result = index.put(document);
+        try {
+            putDocumentWithRetry(indexName, document);
+        } catch (TeammatesException e) {
+            log.severe(String.format(ERROR_NON_TRANSIENT_BACKEND_ISSUE, document, indexName)
+                    + TeammatesException.toStringWithStackTrace(e));
+        } catch (MaximumRetriesExceededException e) {
+            log.severe(String.format(ERROR_MAXIMUM_RETRIES_EXCEEDED, document, indexName)
+                    + TeammatesException.toStringWithStackTrace(e));
+        }
+    }
 
-                if (Config.PERSISTENCE_CHECK_DURATION == 0) {
-                    continue;
+    /**
+     * Tries putting a document, handling transient errors by retrying with exponential backoff.
+     *
+     * @throws TeammatesException if a non-transient error is encountered.
+     * @throws MaximumRetriesExceededException with final {@link OperationResult} if operation fails after maximum retries.
+     */
+    private static void putDocumentWithRetry(String indexName, final Document document)
+            throws TeammatesException, MaximumRetriesExceededException {
+        final Index index = getIndex(indexName);
+
+        /**
+         * The GAE Search API signals put document failure in two ways: it either
+         * returns a {@link PutResponse} containing an {@link OperationResult} with a non-OK {@link StatusCode}, or
+         * throws a {@link PutException} that also contains an embedded {@link OperationResult}.
+         * We handle both ways by examining the {@link OperationResult} to determine what kind of error it is. If it is
+         * transient, we use {@link RetryManager} to retry the operation; if it is
+         * non-transient, we do not retry but throw a {@link TeammatesException} upwards instead.
+         */
+        RM.runUntilSuccessful(new RetryableTaskReturnsThrows<OperationResult, TeammatesException>("Put document") {
```

Comment:
Not sure if `TeammatesException` should be used here as it seems to denote some sort of logical error in the system, rather than failure with datastore or GAE.